stock_archive_constraint/models/product_product.py

```diff
@@ -0,0 +1,44 @@
+# Copyright 2020 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    @api.constrains("active")
+    def _check_active_stock_archive_constraint(self):
+        res = self.env['stock.quant'].search(
+            [
+                ('product_id', 'in', self.ids),
+                ('quantity', '!=', 0.0)
+            ], limit=1
+        )
+        if res:
+            raise ValidationError(
+                _("It is not possible to archive a product that has quant associated.")
```

Comment:
Incorrect English text + you can be archiving several in batch and you need to know which one:
```suggestion
                _("It is not possible to archive product '%s' which has associated quants.", res.product_id.display_name)
```